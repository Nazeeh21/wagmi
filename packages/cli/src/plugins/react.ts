import { pascalCase } from 'change-case'
import dedent from 'dedent'

import type { Plugin } from '../config'

type ReactConfig = {
  hooks?: {
    useContract?: boolean
    useContractEvent?: boolean
    useContractRead?: boolean
    useContractWrite?: boolean
    usePrepareContractWrite?: boolean
  }
}

export function react(config: ReactConfig): Plugin {
  const hooks = {
    useContract: true,
    useContractEvent: true,
    useContractRead: true,
    useContractWrite: true,
    usePrepareContractWrite: true,
    ...config?.hooks,
  }
  return {
    name: 'React',
    async run({ contracts, content, isTypeScript }) {
      const imports = new Set<string>([])
      for (const contract of contracts) {
        const hasAddress = !!contract.address
        const hasMultichainAddress = typeof contract.address === 'object'

        let addressConfig, addressType
        if (hasMultichainAddress) {
          addressConfig = `${contract.meta.addressName}[config.chainId]`
          addressType = `chainId: keyof typeof ${contract.meta.addressName}`
        } else if (hasAddress) {
          addressConfig = contract.meta.addressName
          addressType = ''
        } else {
          addressType = `address?: UseContractConfig['address']`
        }
        const contractConfig = dedent`
          abi: ${contract.meta.abiName},
          ${addressConfig ? `address: ${addressConfig},` : ''}
        `

        let hasReadFunction,
          hasWriteFunction,
          hasEvent = false
        for (const component of contract.abi) {
          if (component.type === 'function')
            if (
              component.stateMutability === 'view' ||
              component.stateMutability === 'pure'
            )
              hasReadFunction = true
            else hasWriteFunction = true
          else if (component.type === 'event') hasEvent = true

          // Exit early if all flags are `true`
          if (hasReadFunction && hasWriteFunction && hasEvent) break
        }

        const content = []
        if (hooks.useContract) {
          imports.add('useContract')
          if (isTypeScript) imports.add('UseContractConfig')
          content.push(
            template.useContract({
              addressType,
              contractConfig,
              isConfigRequired: hasMultichainAddress,
              isTypeScript,
              name: contract.name,
            }),
          )
        }
        if (hasReadFunction) {
          if (hooks.useContractRead) {
            imports.add('useContractRead')
            if (isTypeScript) imports.add('UseContractReadConfig')
            if (hasReadFunction) {
              content.push(
                template.useContractRead({
                  addressType,
                  contractConfig,
                  isConfigRequired: hasMultichainAddress,
                  isTypeScript,
                  name: contract.name,
                }),
              )
            }
          }
        }

        contract.content = `${contract.content}\n\n${content.join('\n\n')}`
      }

      return {
        contracts,
        content: {
          header: [...content.header],
          imports: [
            ...content.imports,
            `import { ${[...imports.values()].join(', ')} } from 'wagmi'`,
          ],
        },
      }
    },
  }
}

type SharedTemplateConfig = {
  addressType: string
  contractConfig: string
  isConfigRequired: boolean
  isTypeScript: boolean
  name: string
}

const template = {
  useContract(config: SharedTemplateConfig) {
    const {
      addressType,
      contractConfig,
      isConfigRequired,
      isTypeScript,
      name,
    } = config
    const hookName = getHookName({ name })

    let typeName = ''
    let typeContent = ''
    if (isTypeScript) {
      typeName = getTypeName({ hookName })
      typeContent = dedent`
        type ${typeName} = Pick<UseContractConfig, 'signerOrProvider'> ${
        addressType
          ? `& {
          ${addressType}
        }`
          : ''
      }
      `
    }

    return dedent`
      ${typeContent}
      export function ${hookName}(config${typeName ? `: ${typeName}` : ''}${
      isConfigRequired ? '' : '= {}'
    }) {
        return useContract({
          ${contractConfig}
          ...config,
        })
      }
    `
  },
  useContractRead(config: SharedTemplateConfig) {
    const {
      addressType,
      contractConfig,
      isConfigRequired,
      isTypeScript,
      name,
    } = config
    const hookName = getHookName({ name, suffix: 'read' })

    let typeName = ''
    let typeContent = ''
    if (isTypeScript) {
      typeName = getTypeName({ hookName })
      typeContent = dedent`
        type ${typeName}<TFunctionName = string> = Omit<
          UseContractReadConfig<typeof wagmiMintEtherscanABI, TFunctionName>,
          ${[`'abi'`, ...(addressType ? [`'address'`] : [])].join(' | ')}
        > ${
          addressType
            ? `& {
            ${addressType}
          }`
            : ''
        }
      `
    }

    return dedent`
      ${typeContent}
      export function ${hookName}(config${typeName ? `: ${typeName}` : ''}${
      isConfigRequired ? '' : '= {}'
    }) {
        return useContractRead({
          ${contractConfig}
          ...config,
        })
      }
    `
  },
}

function getHookName({ name, suffix }: { name: string; suffix?: string }) {
  return `use${pascalCase(name)}${suffix ? pascalCase(suffix) : ''}`
}

function getTypeName({ hookName }: { hookName: string }) {
  return `${pascalCase(hookName)}Config`
}
