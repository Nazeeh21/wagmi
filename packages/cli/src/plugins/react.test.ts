import fixtures from 'fixturez'
import { describe, expect, it, vi } from 'vitest'

import { react } from './react'

const f = fixtures(__dirname)

describe('react', () => {
  it('validate', async () => {
    const temp = f.temp()
    const spy = vi.spyOn(process, 'cwd')
    spy.mockImplementation(() => temp)

    await expect(react().validate()).rejects
      .toThrowErrorMatchingInlineSnapshot(`
      "wagmi must be installed to use React plugin.
      To install, run: pnpm add wagmi"
    `)

    vi.restoreAllMocks()
  })

  describe('run', () => {
    describe('isTypeScript', () => {
      const abi = [
        {
          inputs: [{ name: 'name', type: 'address' }],
          name: 'balanceOf',
          outputs: [{ name: 'balance', type: 'uint256' }],
          stateMutability: 'view',
          type: 'function',
        },
      ] as const

      it('true', async () => {
        await expect(
          react().run({
            contracts: [
              {
                name: 'Wagmi',
                address: '0xaf0326d92b97df1221759476b072abfd8084f9be',
                abi,
                content: '',
                meta: {
                  abiName: 'WagmiAbi',
                  addressName: 'WagmiAddress',
                  configName: 'WagmiConfig',
                },
              },
            ],
            isTypeScript: true,
          }),
        ).resolves.toMatchSnapshot()
      })

      it('false', async () => {
        await expect(
          react().run({
            contracts: [
              {
                name: 'Wagmi',
                address: '0xaf0326d92b97df1221759476b072abfd8084f9be',
                abi,
                content: '',
                meta: {
                  abiName: 'WagmiAbi',
                  addressName: 'WagmiAddress',
                  configName: 'WagmiConfig',
                },
              },
            ],
            isTypeScript: false,
          }),
        ).resolves.toMatchSnapshot()
      })
    })
  })
})
