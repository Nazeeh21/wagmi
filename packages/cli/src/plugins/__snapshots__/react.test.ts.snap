// Vitest Snapshot v1

exports[`react > run > isTypeScript > false 1`] = `
{
  "content": "/**
* Wraps {@link useContract} with \`abi\` set to {@link WagmiAbi}. 
*/
   export function useWagmi(config = {}) {
     return useContract({
abi: WagmiAbi,
address: WagmiAddress,...config})
   }

/**
* Wraps {@link useContractRead} with \`abi\` set to {@link WagmiAbi}. 
*/
export function useWagmiRead(config = {}) {
  return useContractRead({
abi: WagmiAbi,
address: WagmiAddress,...config})
}

/**
* Wraps {@link useContractRead} with \`abi\` set to {@link WagmiAbi} and \`functionName\` set to \`\\"balanceOf\\"\`. 
*/
export function useWagmiBalanceOf(config = {}) {
  return useContractRead({
abi: WagmiAbi,
address: WagmiAddress,
functionName: \\"balanceOf\\",...config})
}",
  "imports": "import { useContract, useContractRead } from 'wagmi'",
}
`;

exports[`react > run > isTypeScript > true 1`] = `
{
  "content": "/**
* Wraps {@link useContract} with \`abi\` set to {@link WagmiAbi}. 
*/
             export function useWagmi(config: Omit<UseContractConfig, \\"abi\\"> = {}as any) {
               return useContract({
abi: WagmiAbi,
address: WagmiAddress,...config})
             }

/**
* Wraps {@link useContractRead} with \`abi\` set to {@link WagmiAbi}. 
*/
export function useWagmiRead<
  TAbi extends readonly unknown[] = typeof WagmiAbi,
  TFunctionName extends string = string
>(
  config: Omit<UseContractReadConfig<TAbi, TFunctionName>, \\"abi\\"> = {}as any,
) {
  return useContractRead({
abi: WagmiAbi,
address: WagmiAddress,...config} as UseContractReadConfig<TAbi, TFunctionName>)
}

/**
* Wraps {@link useContractRead} with \`abi\` set to {@link WagmiAbi} and \`functionName\` set to \`\\"balanceOf\\"\`. 
*/
export function useWagmiBalanceOf<
  TAbi extends readonly unknown[] = typeof WagmiAbi,
  TFunctionName extends string = \\"balanceOf\\"
>(
  config: Omit<UseContractReadConfig<TAbi, TFunctionName>, \\"abi\\" | \\"functionName\\"> = {}as any,
) {
  return useContractRead({
abi: WagmiAbi,
address: WagmiAddress,
functionName: \\"balanceOf\\",...config} as UseContractReadConfig<TAbi, TFunctionName>)
}",
  "imports": "import { useContract, UseContractConfig, useContractRead, UseContractReadConfig } from 'wagmi'",
}
`;
